### ISSUE TITLE
ResponseValidationFailure not reported on server logs

### ISSUE THREAD

**Issue by jeromekelleher** [ga4gh/server#213]
When a response validation occurs, this is transformed into a frontend.ServerException and we must put in print statements to find out what went wrong. This is not very helpful for developers. We should report the type of exception that has occured, and trace back to where it happened.

In general, I struggle to see the purpose of the split between backend and frontend exceptions. What's wrong with all modules just importing `exceptions`, and having a single exception hierarchy? (Sure we can have branches in this hierachy to represenent things that go wrong on the frontend and the backend.) To me this doesn't introduce any real coupling, and does away with a lot of the complicated dancing around we have to do when processing exceptions at the frontend.

**Comment by dcolligan** [comment]
Doesn't the `print(traceback.format_exc(exception))` line in `frontend.py` take care of this?

I agree that unifying the exceptions files would simplify logic.  The split I was attempting was that frontend exceptions need to worry about how they are going to appear to the client (defining httpStatus and an errorCode) while the backend exceptions don't need to worry about that.  If we unified them, what are now the backend exceptions would need to have their httpStatus/code encoded into the class instead of being mapped to a frontend exception as they are now.

**Comment by jeromekelleher** [comment]
No, it doesn't seem to catch it. Try the code in the current master with an 'all variants' query. 

Good point about the httpStatus and errorCode - I'm sure if we thought about it we'd come up with a cunning class hierarchy that sorted out these details. How about this:
```
ServerException
    + FrontendException
    + BackendException
        + RequestValidationException
        + ResponseValidationException
        + ...
```
BackendException has httpStatus=500, and everything below this inherits this httpStatus. Children of FrontendExceptions are pretty much as they are now. All subclasses of ServerException have an automatically generated `errorCode` value which is derived from the exception name (say, the integer encoding of the hash of the fully qualified class name). Does this make sense to you?

**Comment by dcolligan** [comment]
Here is the error that is thrown in the scenario you mentioned:
```
Traceback (most recent call last):
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/dannyc/ga4gh/server/ga4gh/frontend.py", line 224, in searchVariants
    version, flask.request, app.backend.searchVariants)
  File "/Users/dannyc/ga4gh/server/ga4gh/frontend.py", line 159, in handleFlaskPostRequest
    return handleHttpPost(flaskRequest, endpoint)
  File "/Users/dannyc/ga4gh/server/ga4gh/frontend.py", line 107, in handleHttpPost
    responseStr = endpoint(request.get_data())
  File "/Users/dannyc/ga4gh/server/ga4gh/backend.py", line 184, in searchVariants
    self.variantsGenerator)
  File "/Users/dannyc/ga4gh/server/ga4gh/backend.py", line 114, in runSearchRequest
    for obj, nextPageToken in objectGenerator(request):
  File "/Users/dannyc/ga4gh/server/ga4gh/backend.py", line 257, in variantsGenerator
    for variant in iterator:
  File "/Users/dannyc/ga4gh/server/ga4gh/datamodel/variants.py", line 387, in getVariants
    "Specifying call set ids is not supported")
NotImplementedError: Specifying call set ids is not supported
```
NotImplementedError is an exception that is already mapped in our hierarchy, so we don't print the stacktrace to the server console.

So I guess the question is: in which cases do you want the traceback to print?  Any time there is any sort of 500 and we are in DEBUG mode?

If you want to rearrange the exception hierarchy / modules / classes, I would file another issue.

**Comment by jeromekelleher** [comment]
No, this is different. Run the same query without the callsets argument.

I'm not suggesting we rearrange the hierarchy here, I was just trying to discuss the options and how we might simplify things and get feedback.

**Comment by pgrosu** [comment]
@jeromekelleher , @dcolligan - You would need to add code for an interactive console which allows for values to be set as the server is running.  A live server can check these values before starting a request, and if they changed will adapt accordingly in responding to errors.  By the way errors can be stored as objects so that they can be inspected at the granularity of interest, besides saving them as text in a log file.

**Comment by dcolligan** [comment]
@jeromekelleher here's the full output of client (note no callsets argument) and server on latest master (68b6310673aede03df821efc778ef881503b2ff5):
```
$ python client_dev.py -O variants-search -V 1000g_2013 -r 1 http://localhost:8000/v0.5.1
ga4gh.client: 500 {
  "errorCode": 11,
  "message": "Not implemented"
}
Traceback (most recent call last):
  File "client_dev.py", line 7, in <module>
    ga4gh.cli.client_main()
  File "/Users/dannyc/ga4gh/server/ga4gh/cli.py", line 829, in client_main
    runner.run()
  File "/Users/dannyc/ga4gh/server/ga4gh/cli.py", line 384, in run
    self._run(self._httpClient.searchVariants, 'id')
  File "/Users/dannyc/ga4gh/server/ga4gh/cli.py", line 343, in _run
    for result in results:
  File "/Users/dannyc/ga4gh/server/ga4gh/client.py", line 145, in runSearchRequest
    'POST', fullUrl, protocolResponseClass, httpData=data)
  File "/Users/dannyc/ga4gh/server/ga4gh/client.py", line 129, in _doRequest
    self._checkStatus(response)
  File "/Users/dannyc/ga4gh/server/ga4gh/client.py", line 94, in _checkStatus
    response.url, response.status_code))
Exception: Url http://localhost:8000/v0.5.1/variants/search had status_code 500
```
```
$ python server_dev.py ga4gh-example-data
 * Running on http://0.0.0.0:8000/ (Press CTRL+C to quit)
 * Restarting with stat
127.0.0.1 - - [09/Mar/2015 09:15:27] "POST /v0.5.1/variants/search HTTP/1.1" 500
```
With the exception printed right when it is caught in `frontend.py`, I get:
```
Traceback (most recent call last):
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/dannyc/ga4gh/bug/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/dannyc/ga4gh/server/ga4gh/frontend.py", line 223, in searchVariants
    version, flask.request, app.backend.searchVariants)
  File "/Users/dannyc/ga4gh/server/ga4gh/frontend.py", line 158, in handleFlaskPostRequest
    return handleHttpPost(flaskRequest, endpoint)
  File "/Users/dannyc/ga4gh/server/ga4gh/frontend.py", line 107, in handleHttpPost
    responseStr = endpoint(request.get_data())
  File "/Users/dannyc/ga4gh/server/ga4gh/backend.py", line 184, in searchVariants
    self.variantsGenerator)
  File "/Users/dannyc/ga4gh/server/ga4gh/backend.py", line 114, in runSearchRequest
    for obj, nextPageToken in objectGenerator(request):
  File "/Users/dannyc/ga4gh/server/ga4gh/backend.py", line 257, in variantsGenerator
    for variant in iterator:
  File "/Users/dannyc/ga4gh/server/ga4gh/datamodel/variants.py", line 387, in getVariants
    "Specifying call set ids is not supported")
NotImplementedError: Specifying call set ids is not supported
```

**Comment by jeromekelleher** [comment]
I think this is yet-another issue. To get the original problem you need to turn on validation. I'm going to close this issue as there are too many problems with the tabix backend to make reproducing it straight forward.
